lib/web/apiserver.go

```diff
@@ -352,6 +352,9 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*RewritingHandler, error) {
 	h.POST("/webapi/trustedcluster", h.WithAuth(h.upsertTrustedClusterHandle))
 	h.DELETE("/webapi/trustedcluster/:name", h.WithAuth(h.deleteTrustedCluster))
 
+	h.GET("/webapi/apps/:fqdn", h.WithAuth(h.getAppFQDN))
+	h.GET("/webapi/apps/:fqdn/:clusterName/:publicAddr", h.WithAuth(h.getAppFQDN))
```

Comment:
since this endpoint is for resolving FQDN, drop the `fqdn` parameter
also, the first endpoint here seems redundant?

Reply:
The `fqdn` parameter is needed, serves as a hint to resolve the actual/safe FQDN. It's the same logic as in the original `createAppSession` handler.

No, both endpoints may get used from the `webapps` app launcher.

Reply:
Can you show an example usage of this endpoint (example URL and response)?
I'm not familiar with how this FQDN resolution works, it just seems odd that it takes an FQDN and returns one too. Maybe the parameter should be named differently, like `appName`?

Reply:
Say there is a Teleport proxy running at `proxy.example.com`.

With the new `getAppFQDN` endpoint, the following would happen in an ordinary, best-case scenario:
1. The user opens an app via the Web UI.
2. A new tab is opened with the app launcher `https://proxy.example.com/web/launch/app.proxy.example.com/cluster-name/app.proxy.example.com` (matching this [URL pattern](https://github.com/gravitational/webapps/blob/88f20440dec42dfce62981aace4c0c5dcd434734/packages/teleport/src/config.ts#L43)).
3. Since the user-provided FQDN shouldn't be blindly trusted, the app launcher invokes `getAppFQDN` to get a known/safe FQDN for the app. In this case `getAppFQDN` would return the same `app.proxy.example.com` which is expected since the web UI populates the app launcher params based on the same data as those used by `getAppFQDN`.
4. The app launcher would then redirect the user to `https://app.proxy.example.com/x-teleport-init` which in turn redirects back to the app launcher with a `state` value etc. etc.

Now, there is another way to invoke the app launcher instead of going directly through Teleport Web UI:
1. The user opens `whatever.proxy.example.com` in his browser.
2. This triggers the app launcher `https://proxy.example.com/web/launch/whatever.proxy.example.com` (only the FQDN param is now present, since the target cluster and the app's public address aren't known at this point).
3. `getAppFQDN` guesses the intended app (from among those running in the root as well as the leaf clusters) on a best-effort basis and either fails or returns the found FQDN. If the app is configured with a public address (here let's say `app.proxy-example.org`), it could be completely unrelated to the FQDN input. 
4. Etc.

From the robustness/security standpoint, `getAppFQDN` serves to prevent infinite redirects and block malicious FQDN inputs (e.g., stealing valid app sessions, phishing). This was already achieved with `CreateAppSession` which also returns the FQDN as part of its output but with the undesirable side effect of creating a new valid app session (confusing audit log entries, a potential attack vector, ...).

Reply:
OK, I think I understand.
Can you change the URL parameter name here from `:fqdn` to `:untrustedFQDN` or `:userFQDN` to hint that it's different from the return FQDN?